launch_ros/launch_ros/descriptions/composable_lifecycle_node.py

```diff
@@ -0,0 +1,121 @@
+# limitations under the License.
+
+"""Module for a description of a ComposableLifecycleNode."""
+
+from typing import List
+from typing import Optional
+
+import launch
+from launch.condition import Condition
+from launch.some_substitutions_type import SomeSubstitutionsType
+from launch.substitution import Substitution
+# from launch.utilities import ensure_argument_type
+from launch.utilities import perform_substitutions
+from launch_ros.parameters_type import Parameters
+from launch_ros.parameters_type import SomeParameters
+from launch_ros.remap_rule_type import RemapRules
+from launch_ros.remap_rule_type import SomeRemapRules
+from launch_ros.utilities import LifecycleEventManager
+
+from .composable_node import ComposableNode
+
+
+class ComposableLifecycleNode(ComposableNode):
+    """Describes a lifecycle node that can be loaded into a container with other nodes."""
+
+    def __init__(
+        self, *,
+        package: SomeSubstitutionsType,
+        plugin: SomeSubstitutionsType,
+        name: Optional[SomeSubstitutionsType] = None,
+        namespace: Optional[SomeSubstitutionsType] = None,
+        parameters: Optional[SomeParameters] = None,
+        autostart: Optional[bool] = False,
+        remappings: Optional[SomeRemapRules] = None,
+        extra_arguments: Optional[SomeParameters] = None,
+        condition: Optional[Condition] = None,
+    ) -> None:
+        """
+        Initialize a ComposableNode description.
+
+        :param package: name of the ROS package the node plugin lives in
+        :param plugin: name of the plugin to be loaded
+        :param name: name to give to the ROS node
+        :param namespace: namespace to give to the ROS node
+        :param parameters: list of either paths to yaml files or dictionaries of parameters
+        :param autostart: Whether to autostart lifecycle node in the activated state
+        :param remappings: list of from/to pairs for remapping names
+        :param extra_arguments: container specific arguments to be passed to the loaded node
+        :param condition: action will be executed if the condition evaluates to true
+        """
+        super().__init__(
+            package=package,
+            plugin=plugin,
+            name=name,
+            namespace=namespace,
+            parameters=parameters,
+            remappings=remappings,
+            extra_arguments=extra_arguments,
+            condition=condition)
+
+        self.__autostart = autostart
+        self.__lifecycle_event_manager = None
+        self.__node_name = self._ComposableNode__node_name
+
+    def init_lifecycle_event_manager(self, node, context: launch.LaunchContext) -> None:
+        # LifecycleEventManager needs a pre-substitution node name
+        self.__node_name = perform_substitutions(context, node.node_name)
+        self.__lifecycle_event_manager = LifecycleEventManager(node)
+        self.__lifecycle_event_manager.setup_lifecycle_manager(context)
```

Comment:
Following up on my previous comment, this could just be:

```suggestion
    def init_lifecycle_event_manager(self, context: launch.LaunchContext) -> None:
        # LifecycleEventManager needs a pre-substitution node name
        self.__lifecycle_event_manager = LifecycleEventManager(self)
        self.__lifecycle_event_manager.setup_lifecycle_manager(context)
```

Also, the `node_name` line seems completely unnecessary? Either that or the line in the constructor of this class where you do `self.__node_name = self._ComposableNode__node_name` isn't needed?

Reply:
Done - actually did this change from the previous comment before I got down to this one :laughing: 

> Also, the node_name line seems completely unnecessary? Either that or the line in the constructor of this class where you do self.__node_name = self._ComposableNode__node_name isn't needed?

It is actually necessary. The node name at the start is a substitution and we need toperform the substitution at this point to get the actual value to pass into the lifecycle event manager.